### PR TITLE
Refactor duplicate code for stringToNumber() in JSONObject, JSONArray, and XML

### LINKED ISSUE #813: Number parsing - de-duplication of logic
Logic to parse number is duplicated in XML and JSON parser.
They can be moved to single place and be referred by both.

### PR DESCRIPTION
**Pull request by rudrajyotib**
Both XML and JSON parser do convert string to number. Similar code has been duplicated for both parsers, which has been moved to a single utility method.

Fixes #813

### CODE REVIEW

File: src/main/java/org/json/NumberConversionUtil.java
Comment (johnjaylward) [inline src/main/java/org/json/NumberConversionUtil.java:6]: Let's not make this public. We may wish to refactor this class in the future. For now make it private or default (package) visible.

**Review by johnjaylward — CHANGES_REQUESTED** [review]

File: src/main/java/org/json/NumberConversionUtil.java
Comment (rudrajyotib) [reply src/main/java/org/json/NumberConversionUtil.java]: Static methods meant to be used by functional classes being marked private do not add up. But yes, @stleary - unnecessary making these classes open for used by clients might not be a great idea, there are plenty of scope of improvements in there. So marked them default visibility.

Comment (johnjaylward) [reply src/main/java/org/json/NumberConversionUtil.java]: That is correct. I simply meant that the class itself should be default or private, not the static methods.

Comment (johnjaylward) [reply src/main/java/org/json/NumberConversionUtil.java]: However, I now recall that private outer classes is not a thing in Java. I was thinking inner class scopes. Default visibility is correct here.

Comment (rudrajyotib) [reply src/main/java/org/json/NumberConversionUtil.java]: @stleary @johnjaylward - can this PR be accepted please?

Comment (stleary) [reply src/main/java/org/json/NumberConversionUtil.java]: @rudrajyotib Will get to it as soon as I can after the work of #811 is completed.

**Comment by rudrajyotib** [comment]
@stleary - can you accept this PR please?

**Comment by stleary** [comment]
@rudrajyotib Please update the PR:

From https://github.com/rudrajyotib/JSON-java
 * branch            issue813   -> FETCH_HEAD
hint: You have divergent branches and need to specify how to reconcile them.
hint: You can do so by running one of the following commands sometime before
hint: your next pull:
hint: 
hint:   git config pull.rebase false  # merge
hint:   git config pull.rebase true   # rebase
hint:   git config pull.ff only       # fast-forward only
hint: 
hint: You can replace "git config" with "git config --global" to set a default
hint: preference for all repositories. You can also pass --rebase, --no-rebase,
hint: or --ff-only on the command line to override the configured default per
hint: invocation.
fatal: Need to specify how to reconcile divergent branches.

File: src/test/java/org/json/NumberConversionUtilTest.java
Comment (Mifrill) [inline src/test/java/org/json/NumberConversionUtilTest.java:59]: ⚙️ alignment
```suggestion
    @Test
```

**Comment by rudrajyotib** [comment]
@stleary - PR is now aligned with HEAD.

**Comment by rudrajyotib** [comment]
@stleary - can you please see if this can be accepted, or is it still showing some discrepancy across git branches.

**Comment by stleary** [comment]
**What problem does this code solve?**
Refactoring duplicate code

**Does the code still compile with Java6?**
Yes

**Risks**
Low

**Changes to the API?**
No

**Will this require a new release?**
No

**Should the documentation be updated?**
No

**Does it break the unit tests?**
No, new unit tests were added

**Was any code refactored in this commit?**
Yes, this was a refactoring ticket

**Review status**
**APPROVED - Pending completion of code review**

Start 3-day comment window

File: src/main/java/org/json/JSONObject.java
Comment (stleary) [inline src/main/java/org/json/JSONObject.java:31]: Just import NumberConversionUtil, not the methods. Someone glancing at the code might not realize it is a static method in another class.

File: src/main/java/org/json/NumberConversionUtil.java
Comment (stleary) [inline src/main/java/org/json/NumberConversionUtil.java:109]: There is a duplicate of this method in JSONObject. It should be removed, and the calling code updated to use the NumberConversionUtil method.

File: src/main/java/org/json/XML.java
Comment (stleary) [inline src/main/java/org/json/XML.java:13]: Just import NumberConversionUtil, not the methods.

File: src/main/java/org/json/XML.java
Comment (stleary) [inline src/main/java/org/json/XML.java:557]: The JSONObject version of this method inserts this line, which is missing from the XML copy:
`initial = val.charAt(0);`. I don't think it matters, though.

**Comment by stleary** [comment]
@rudrajyotib  Looks good to me. I left some comments in the code. They are all trivial issues that can wait until this commit is merged. Please do not make any more changes to this PR; if you do, the accept status will be taken down and it won't get committed before the end of Hacktoberfest. I will open an issue to address the notes after Nov 1, but they are low priority.

**Comment by rudrajyotib** [comment]
@stleary sure thing. I will update them in another PR.

File: src/main/java/org/json/XML.java
Comment (rudrajyotib) [reply src/main/java/org/json/XML.java:557]: This line is not needed in any of the versions, and has been deleted from either, moved into the utility static method.

File: src/main/java/org/json/XML.java
Comment (rudrajyotib) [reply src/main/java/org/json/XML.java:13]: In the next PR it would be changed.

File: src/main/java/org/json/NumberConversionUtil.java
Comment (rudrajyotib) [reply src/main/java/org/json/NumberConversionUtil.java:109]: Right, the duplicate method will be removed in the next PR.

**Review by stleary — APPROVED** [review]

**Pull request merged**
